src/libertem/executor/pipelined.py

```diff
@@ -816,17 +942,33 @@ def yield_result_if_found(block, timeout):
                         result_task_id = result["task_id"]
                         yield (result["result"], id_to_task[result_task_id], result_task_id)
                         del id_to_task[result_task_id]
-                        assert len(id_to_task) == in_flight[0]
+                        if len(id_to_task) != in_flight[0]:
+                            raise RuntimeError(
+                                "state mismatch; `id_to_task` mapping should match `in_flight`"
+                            )
                 except WorkerQueueEmpty:
                     self._validate_worker_state()
 
             for task_idx, task in enumerate(tasks):
+                # important: call `schedule_task` before incrementing number of
+                # in-flight tasks, otherwise, if we run into an exception, we
+                # will wait for a response that never comes.
+                _worker_idx, worker_queues = schedule_task(
```

Comment:
It's just naming, but `schedule_task` makes me think that this function call actually schedules the task, instead it performs a function more aligned with a name like `choose_worker`. The actual scheduling happens when we send `RUN_TASK` (for me).

Reply:
Naming is hard ;) For me, scheduling is more of a planning step, but I can understand that this can be a bit confusing. `choose_worker` IMHO also doesn't fully fit, as it is a different step than the `selector` function that is passed in (but includes calling said `selector`).. Hmm, I'll leave it as-is for now, until we can agree on/find a better name.

Reply:
No worries, could just add a comment above the call saying that we are selecting the most appropriate worker for the task here.